transformer_lens/HookedSAE.py

```diff
@@ -0,0 +1,115 @@
+            torch.nn.init.kaiming_uniform_(
+                torch.empty(self.cfg.d_sae, self.cfg.d_in, dtype=self.cfg.dtype)
+            )
+        )
+        self.b_enc = nn.Parameter(torch.zeros(self.cfg.d_sae, dtype=self.cfg.dtype))
+        self.b_dec = nn.Parameter(torch.zeros(self.cfg.d_in, dtype=self.cfg.dtype))
+
+        self.hook_sae_input = HookPoint()
+        self.hook_sae_acts_pre = HookPoint()
+        self.hook_sae_acts_post = HookPoint()
+        self.hook_sae_recons = HookPoint()
+        self.hook_sae_error = HookPoint()
+        self.hook_sae_output = HookPoint()
+
+        self.to(self.cfg.device)
+        self.setup()
+
+    def forward(self, input: Float[torch.Tensor, "... d_in"]) -> Float[torch.Tensor, "... d_in"]:
+        """SAE Forward Pass.
+
+        Args:
+            input: The input tensor of activations to the SAE. Shape [..., d_in].
+                Also supports hook_z activations of shape [..., n_heads, d_head], where n_heads * d_head = d_in, for attention output (hook_z) SAEs.
+
+        Returns:
+            output: The reconstructed output tensor from the SAE, with the error term optionally added. Same shape as input (eg [..., d_in])
+        """
+        self.hook_sae_input(input)
+        if input.shape[-1] == self.cfg.d_in:
+            x = input
+        else:
+            # Assume this this is an attention output (hook_z) SAE
+            assert self.cfg.hook_name.endswith(
+                "_z"
+            ), f"You passed in an input shape {input.shape} does not match SAE input size {self.cfg.d_in} for hook_name {self.cfg.hook_name}. This is only supported for attn output (hook_z) SAEs."
+            x = einops.rearrange(input, "... n_heads d_head -> ... (n_heads d_head)")
+        assert (
+            x.shape[-1] == self.cfg.d_in
+        ), f"Input shape {x.shape} does not match SAE input size {self.cfg.d_in}"
+
+        x_cent = x - self.b_dec
+        sae_acts_pre = self.hook_sae_acts_pre(
+            einops.einsum(x_cent, self.W_enc, "... d_in, d_in d_sae -> ... d_sae")
+            + self.b_enc  # [..., d_sae]
+        )
+        sae_acts_post = self.hook_sae_acts_post(F.relu(sae_acts_pre))  # [..., d_sae]
+        x_reconstruct = self.hook_sae_recons(
+            (
+                einops.einsum(sae_acts_post, self.W_dec, "... d_sae, d_sae d_in -> ... d_in")
+                + self.b_dec
+            ).reshape(input.shape)
+        )
+
+        if self.cfg.use_error_term:
+            with torch.no_grad():
+                # Recompute everything without hooks to get true error term
+                # Otherwise, the output with error term will always equal input, even for causal interventions that affect x_reconstruct
+                # This is in a no_grad context to detach the error, so we can compute SAE feature gradients (eg for attribution patching). See A.3 in https://arxiv.org/pdf/2403.19647.pdf for more detail
+                sae_acts_pre_clean = (
```

Comment:
```suggestion
                # NOTE: we can't just use `sae_error = input - x_reconstruct.detach()` or something simpler, since this would mean intervening on features would mean ablating features still results in perfect reconstruction.
                sae_acts_pre_clean = (
```